### PR TITLE
[MRG] New DQN implementation

### PR DESCRIPTION
**Pull request by omardrwch**
New implementation of `agents.torch.DQNAgent`:

* New replay buffer class that does not use `np.random` directly (to fix reproducibility issues);
* Possibility of using Q(lambda);
* Improve docs for DQN and torch utils.

### CODE REVIEW

File: rlberry/tests/test_agent.py [hunk 1]
```diff
@@ -46,7 +46,7 @@ def feature_map_fn(_env):
 CONTINUOUS_STATE_AGENTS = [
     agents.RSUCBVIAgent,
     agents.RSKernelUCBVIAgent,
-    # torch_agents.DQNAgent,  # For now, DQN does not work with the generic check.
+    torch_agents.DQNAgent,
```
Comment (TimotheeMathieu) [inline rlberry/tests/test_agent.py:49]: Now that DQNAgent pass the checkes, you can remove the custom check I made line 66 as it has become useless.

File: rlberry/agents/torch/dqn/dqn.py [hunk 2]
```diff
@@ -1,568 +1,449 @@
...
-    batch_size : int
-        Batch size
-    device : str
-        Device used by pytorch.
-    target_update : int
-        Number of steps to wait before updating the target network.
-    double : bool
-        If true, use double Q-learning.
-    learning_rate : double
+    env: :class:`rlberry.types.Env`
```
Comment (TimotheeMathieu) [inline rlberry/agents/torch/dqn/dqn.py:48]: I think you missed a tilde :class:`~rlberry.types.Env` in order to make the link.

File: rlberry/agents/torch/dqn/dqn.py [hunk 3]
```diff
@@ -1,568 +1,449 @@
...
-            )
-            model_input = torch.zeros(
-                (1, *obs_shape), dtype=torch.float, device=self.device
-            )
-            self.writer.add_graph(self.value_net, input_to_model=(model_input,))
-            self.writer.add_scalar(
-                "agent/trainable_parameters", trainable_parameters(self.value_net), 0
-            )
+if __name__ == "__main__":
+    from rlberry.envs import gym_make
```
Comment (TimotheeMathieu) [inline rlberry/agents/torch/dqn/dqn.py:442]: I don't really understand why you do this. I would expect this kind of example to go into the example folder, I think that usually we don't include example in the library source. If you really want to provide an example you can put it in the docstring as an example but I think this precise example would be too ling to run (docstring examples are automatically run when the doc is constructed to check that it compiles).

**Review by TimotheeMathieu — COMMENTED** [review]
I don't have a lot to say because I am not really competent to review a DQN but I still have a few comments. 
Also, good job on the docstrings, this is very nice and very easy to navigate/undestand.

File: rlberry/agents/torch/dqn/dqn.py (hunk 3 above)
Comment (omardrwch) [reply rlberry/agents/torch/dqn/dqn.py:442]: Indeed, I'm going to remove it, I was only using it to debug the implementation and forgot to delete this part!

**Review by mmcenta — COMMENTED** [review]
I'm OK with the changes - great coding style and documentation.

I made a few notes on small style things, but there are two main things that I wanted to note about the replay buffer:

1. When you correct a `bad starting index` to the current position minus the chunk size, you're changing the distribution by concentrating probability on the most recent chunk. Ideally, you should resample bad indices but I'm not sure that this matters that much.
2. Are chunks (or sub-trajectories) supposed to be part of the same episode? Because I don't think there's anything preventing the replay buffer from returning a chunk that contains an environment reset.

File: rlberry/agents/utils/replay.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+        """
+        assert chunk_size > 0
+        assert sampling_mode in ["uniform", "prioritized"]
+        if sampling_mode == "prioritized" and not self._enable_prioritized:
+            raise RuntimeError(
+                "To sample from the replay with the 'prioritized' mode, "
+                "you need to set enable_prioritized=True in the constructor."
+            )
+
+        if len(self) <= chunk_size:
```
[the remark this answers is not in the document]
Comment (omardrwch) [reply rlberry/agents/utils/replay.py:290]: I think throwing an exception might be bad if the user code has a loop waiting for the replay buffer to be filled. So I'm adding more docs about what is returned, and I'm logging an error message in this case.

File: rlberry/agents/tests/test_replay.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    )
+    buffer.setup_entry("observations", np.float32)
+    buffer.setup_entry("actions", np.uint32)
+    buffer.setup_entry("rewards", np.float32)
+    buffer.setup_entry("dones", np.bool)
+
+    # Fill the replay buffer
+    total_time = 0
+    while True:
+        if total_time > 2 * buffer._max_replay_size:
```
[the remark this answers is not in the document]
Comment (omardrwch) [reply rlberry/agents/tests/test_replay.py:25]: I wanted to test that the replay size is respecting the constraint of not storing more than `max_replay_size` transitions:
```python
buffer, _ = _get_filled_replay(max_replay_size=500)
rng = buffer._rng
assert len(buffer) == 500  # it should be 500, although we called .append() more than 500 times
```

File: rlberry/agents/torch/dqn/dqn_utils.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    def schedule(count):
+        count = np.clip(count - transition_begin, 0, transition_steps)
+        frac = 1 - count / transition_steps
+        return (init_value - end_value) * (frac**power) + end_value
+
+    return schedule
+
+
+@numba_jit
```
[the remark this answers is not in the document]
Comment (omardrwch) [reply rlberry/agents/torch/dqn/dqn_utils.py:75]: Could you share/link the code using `scipy.signal.lfilter`?  It might be a good idea to use `scipy.signal.lfilter` here, but I think the advantage of numba + loop is that it can be more readable (e.g. seeing clearly what happens when lambda_ = 1 or 0) and there should be no loss of performance.

File: rlberry/agents/torch/dqn/dqn.py [hunk 7]
```diff
@@ -1,568 +1,438 @@
...
-        Use prioritized replay.
+    loss_function: {"l1", "l2", "smooth_l1"}, default: "l2"
+        Loss function used to compute Bellman error.
+    epsilon_init: float, default = 1.0
+        Initial epsilon value for epsilon-greedy exploration.
+    epsilon_final: float, default = 0.1
+        Final epsilon value for epsilon-greedy exploration.
+    epsilon_decay_interval : int
+        After :code:`epsilon_decay` timesteps, epsilon approaches :code:`epsilon_final`.
+    optimizer_type : {"ADAM", "RMS_PROP"}
```
[the remark this answers is not in the document]
Comment (omardrwch) [reply rlberry/agents/torch/dqn/dqn.py:73]: Agreed, I followed what is used by `optimizer_factory`. We have uppercase "ADAM" etc in other agents, so it's better to fix it in another PR. I'll include it in the TODOs.

File: rlberry/agents/torch/dqn/dqn.py [hunk 8]
```diff
@@ -1,568 +1,438 @@
...
-    memory_capacity : int
-        Capacity of the replay buffer (in number of transitions).
-    use_bonus : bool, default = False
-        If true, compute an 'exploration_bonus' and add it to the reward.
-        See also UncertaintyEstimatorWrapper.
-    uncertainty_estimator_kwargs : dict
-        Arguments for the UncertaintyEstimatorWrapper
-    prioritized_replay: bool
-        Use prioritized replay.
+    loss_function: {"l1", "l2", "smooth_l1"}, default: "l2"
```
[the remark this answers is not in the document]
Comment (omardrwch) [reply rlberry/agents/torch/dqn/dqn.py:65]: I think DQN is almost always implemented with "l2" or "smooth_l1", so I'd say we can keep it like this for now. Have you seen another loss that might be useful to include here?

[the remark this answers is not in the document]
Comment (mmcenta) [reply rlberry/agents/torch/dqn/dqn.py:65]: I think a few extensions can be implemented by directly tweaking the loss function, such as algorithms for Offline RL and exploration. I was thinking about those, but I don't know if they can be implemented by just changing the loss.

File: rlberry/agents/torch/dqn/dqn.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mmcenta) [reply rlberry/agents/torch/dqn/dqn.py:73]: That works!

File: rlberry/agents/utils/replay.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mmcenta) [reply rlberry/agents/utils/replay.py:306]: You're right! Thant sound like a good solution.

File: rlberry/agents/torch/dqn/dqn_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mmcenta) [reply rlberry/agents/torch/dqn/dqn_utils.py:75]: You're right. The `signal.lfilter` approach is famously opaque to the user - maybe it is worth it to support `numba` so we can have more readable and flexible code.

On another note, why do we use `numba` for JIT compilation? Is it a requirement for the base version of `rlberry`?

File: rlberry/agents/tests/test_replay.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (omardrwch) [reply rlberry/agents/tests/test_replay.py:29]: I tried to split the tests a bit more now! (one for size, one for sampling, one for priority updates and another to check valid indices)

[the remark this answers is not in the document]
Comment (TimotheeMathieu) [reply rlberry/agents/tests/test_replay.py:29]: If you don't want to make multiple small tests you can use multiple assert, this is how I usually do things with pytest.
Idealy there should be one assert for each thing that can go wrong with an error message for each  assert. Example
```
buffer, _ = _get_filled_replay(max_replay_size=500)
rng = buffer._rng
assert len(buffer) == 500, "length buffer invalid, please check..."
```
then the first false assert will raise an error with your custom error message (here "length buffer invalid, please check..."). You can also use the try, except syntax to raise error messages if you don't know what to assert (as I did in my check utils)

File: rlberry/agents/torch/dqn/dqn_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (omardrwch) [reply rlberry/agents/torch/dqn/dqn_utils.py:75]: Yes, numba is included in the default installation mode. If numba is not installed, the code still runs without JIT compilation (I think readthedocs was not working with numba at some point, so we made it optional, that's why we don't import numba directly).

Numba is nice because it only requires us to add a decorator to a function, and we don't need to use another interpreter (as PyPy), or to change the code (as cython).

File: rlberry/agents/utils/replay.py [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+        # Handle invalid sub-trajectories: those containing a transition
+        # corresponding to self._position. That is, we need to handle
+        # the fact that we're sampling sub-trajectories from a circular buffer
+        # and that we cannot cross the 'boundary' defined by self._position.
+        #
+        # What we do here is to move the start_index to the most recent
+        # index (with respect to self._position) that allows sampling
+        # a sub-trajectory of length chunk_size.
+        # Note that this can result in a negative start_index,
+        # which is handled in the _get_chunk method.
```
[the remark this answers is not in the document]
Comment (omardrwch) [reply rlberry/agents/utils/replay.py:240]: Indeed, that's the trickiest part of this implementation. I thought about 2., but it would make the runtime stochastic, which I don't think is desirable. We're indeed biasing the distribution, but the bias decreases as the size of the buffer increases (and no bias for chunk_size=1), so I think the option 1. (keep it like this) is ok.

I haven't given it much thought, but I couldn't immediately figure out a simple way of sampling over a distribution over chunks in O(batch_size * chunk_size) deterministic time.

I'd say (little bias + no rejection sampling) is better than (no bias + rejection sampling). What do you think?

[the remark this answers is not in the document]
Comment (mmcenta) [reply rlberry/agents/utils/replay.py:240]: I think that if you could compute the chunk probabilities in the prioritized case by making a few adjustments to the `_sample_proportional_and_compute_weights` function:

1. Instead of setting `p_total = it_sum.sum(0, len(self) - 1)` sum only over the valid indices;
2. At each range, instead of using the sampled `mass` directly, call `find_prefixsum_idx` with the `mass` plus the mass of the indices we're ignoring.

This would look like something like this:

```python
# Valid indices are part of the range [start_idx, end_idx]
start_mass, end_mass = self._it_sum.sum(0, start_idx), self._it_sum.sum(0, end_idx)

# Computes the sum of priorities from start_idx to end_idx
if start_mass <= end_mass:
  p_total = self._it_sum.sum(0, len(self)-1) - end_mass + start_mass
else:
  p_total = start_mass - end_mass

# When computing the index from a sample, we must add the ignored mass again
mass = ...
if start_mass <= end_mass:
  mass += start_mass
elif mass > start_mass:
  mass += start_mass - end_mass
idx = self._it_sum.find_prefixsum_idx(mass)
```

[the remark this answers is not in the document]
Comment (mmcenta) [reply rlberry/agents/utils/replay.py:240]: Of course, this method also introduces some bias because it sets the chunk priority to the priority of the first element of the chunk (one could argue that maybe choosing the last element is better). I think any of these implementations (including your original) is fine, but it would be interesting to test if these choices matter.

[the remark this answers is not in the document]
Comment (omardrwch) [reply rlberry/agents/utils/replay.py]: I just pushed a commit to remove the bias from the `"uniform"` sampling mode. For prioritized replay, I think it's better to keep it like this for now: it's simpler and the bias only increases a little the probability of sampling the last chunk; and, in prioritized experience replay, the last transitions are initialized with the highest priority anyway. So I think the impact of this choice shouldn't matter too much.

**Review by mmcenta — APPROVED** [review]
LGTM
